### PR TITLE
Adding api builder for cloudstack machine templates and machine deployments

### PR DESCRIPTION
**Pull request by maxdrib**
Issue #, if available:
https://github.com/aws/eks-anywhere/issues/2260

Description of changes:
First PR of many covering changes required to support full lifecycle API controller. This moves the cloudstackmachinetemplate generation for control planes to use the apibuilder instead of templating.

Testing (if applicable):
* unit tests
* successfully ran 1.20->1.21 E2E test locally
* Successfully scaled cluster with changes from https://github.com/aws/eks-anywhere/pull/2938

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/cloudstack/apibuilder.go
Comment (vivek-koppuru) [inline pkg/providers/cloudstack/apibuilder.go:22]: Can we use the `InfrastructureAPIVersion` here from `cloudstackv1` or is there no method that gets this specific version?

File: pkg/providers/cloudstack/apibuilder.go
Comment (vivek-koppuru) [inline pkg/providers/cloudstack/apibuilder.go:33]: This logic can be unit tested right?

File: pkg/providers/cloudstack/apibuilder.go
Comment (vivek-koppuru) [inline pkg/providers/cloudstack/apibuilder.go:57]: Instead of needing to sort in the business logic, can we just change our verification logic in our unit test to use a hash set?

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:22]: Indeed we can, thanks

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:57]: The unit tests in provider/cloudstack.go are strictly validating against yaml "expected" testdata at the moment which don't have this capability :(

File: pkg/providers/cloudstack/cloudstack.go
Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:662]: You are correct, thanks for the note

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:33]: Required some additional setup but it's possible!

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [inline pkg/providers/cloudstack/apibuilder.go:81]: Maybe we should have the name being generated in this method?

File: pkg/providers/cloudstack/cloudstack.go
Comment (g-gaston) [inline pkg/providers/cloudstack/cloudstack.go:666]: I would like the machine template name to come from a method, so it's easy to replicate

Comment (g-gaston) [inline pkg/providers/cloudstack/cloudstack.go:666]: Just so I can see how this will fit the big picture, how are you planning on using this from the controller?

Are the code paths gonna be the same for cli and controller (minus the marshaling) or are they going to be composed differently

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: I tried not including this field but CAPC appeared to reject it with some validation webhook error saying `CreationTimestamp cannot be null` or something along those lines. I don't believe it's actually used anywhere in CAPC, but it has to be set.

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:81]: Name generation is complex, and depends on the usecase. It varies between the CLI and controller, and between old controller and new controller. I think it would be cleaner to leave that logic upstream and keep this method simple.

File: pkg/providers/cloudstack/cloudstack.go
Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:666]: all the immutable objects are going to be generated from the apibuilder (machinetemplates and machinedeployments) in both the CLI and the controller. All the mutable objects will be generated from the templates (kubeadmconfig, cloudstackcluster, capi cluster) in both the CLI and controller. In the CLI, the objects will be applied as yaml. In the controller, they will be compared and then applied as K8s objects.

Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:666]: currently, machine template name is only available from the `values` map. All this code is doing is reading from that map and then converting to a string. Are you suggesting factoring out that logic? Or are you proposing we centralize all the ways that the name is generated?

File: pkg/providers/cloudstack/apibuilder.go
Comment (vivek-koppuru) [reply pkg/providers/cloudstack/apibuilder.go:56]: Can we create an action item to address this? I would really like to avoid doing something for a unit test in our business logic. Not sure if we do this today already, but ideally the symlinks passed into this method would already be sorted for that case. Are we able to control that?

File: pkg/providers/cloudstack/apibuilder.go
Comment (vivek-koppuru) [reply pkg/providers/cloudstack/apibuilder.go:96]: If it can't be null, then can we set it to an empty struct and have the system fill it in still? Or is that not possible?

File: pkg/providers/cloudstack/cloudstack.go
Comment (vivek-koppuru) [reply pkg/providers/cloudstack/cloudstack.go:662]: I think `cpTemplateName` or `cpMachineTemplateName` is fine to say here, no need to say `Cs`, unless we have a reference to the regular machine template name here that is different than this.

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:56]: The symlinks are passed in as a standard unsorted hashmap and they get sorted arbitrarily when you iterate over them. The other option would be only testing for a single symlink but then we do not get validation on the multi-symlink formatting, which feels like missed coverage and risky. I agree about adding logic for unit tests to business logic is messy, but it seems harmless in any real scenario in my opinion so I'm inclined to keep it for now.

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:96]: @maxdrib do you think we can track where that error comes from? I'm ok moving forward with this if it's a blocker but it sounds weird that webhook will reject this. Looking upstream, it says its set by the system:
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L179-L188

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:81]: I would expect the CLI and the new controller to use the same name generation logic. How do those two differ right now?

File: pkg/providers/cloudstack/cloudstack.go
Comment (g-gaston) [reply pkg/providers/cloudstack/cloudstack.go:666]: I see. But we are in `GenerateCAPISpecControlPlane` method, which returns `[]byte`
Then does this mean we are going to need to duplicate the logic that calls "go template + api structs" in the controller? Just to be able to access the api structs before they get marshaled to yaml

Comment (g-gaston) [reply pkg/providers/cloudstack/cloudstack.go:666]: I see. But we are in `GenerateCAPISpecControlPlane` method, which returns `[]byte`
Then does this mean we are going to need to duplicate the logic that calls "go template + api structs" in the controller? Just to be able to access the api structs before they get marshaled to yaml

Comment (g-gaston) [reply pkg/providers/cloudstack/cloudstack.go:666]: Ah I see
Is the machine template name being used in the go template? If not, no need to pass it as part of the map, we can make it a explicit argument to this/other methods

As for conversion, does `fmt.Sprintf("%s", cpCsMtName)` work?

File: pkg/providers/cloudstack/apibuilder.go
Comment (chrisdoherty4) [reply pkg/providers/cloudstack/apibuilder.go:96]: Is it getting rejected by a CAPC webhook?

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: Correct @chrisdoherty4, it is being rejected by a CAPC webhook when it's not set. I don't remember the exact logs. I could set this to an empty Time object but that'll be misleading since it'll be epoch time I imagine.

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]:
```
2022-09-26T14:59:10.137-0400	V6	Executing command	{"cmd": "/usr/local/bin/docker exec -i eksa_1664218435037679000 kubectl apply -f - --namespace eksa-system --kubeconfig eksa-drib-b77a195/generated/eksa-drib-b77a195.kind.kubeconfig"}
2022-09-26T14:59:10.832-0400	V9	docker	{"stderr": "error: error validating \"STDIN\": error validating data: unknown object type \"nil\" in CloudStackMachineTemplate.spec.template.metadata.creationTimestamp; if you choose to ignore these errors, turn validation off with --validate=false\n"}
```

Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:96]: that doesn't seem to come from the webhook necessarily, it seems more like the api server
Is this a new machine template or are you updating an existing one?
If it's an update and you set the date during creation, that date would be part of the last-applied annotation, hence kubectl would try to remove the date that was already set.

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: It's a new one - this error appears on cluster creation when I exclude the ObjectMeta entirely. I think previously I had seen some actual webhook error from CAPC but I can't find the log now, it's been a few days.

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: This seems to be a serialization issue. For some reason, no matter what I try, it's being serialized as 
```
    metadata:
      creationTimestamp: null
```
which the api server does not like for some reason. This is also how the creationTimestamp appears in the CSMT's top level ObjectMeta (`creationTimestamp: null`) even when it's not explicitly specified. 

When we were using template-cp, the resulting spec did not have ObjectMeta specified here, and CAPC was happy with that. With that being said, I think we have three options to proceed here

1. Remove this ObjectMeta attribute from the CloudStackMachineTemplate.Spec.Spec in CAPC. This will require an API bump in CAPC and does not feel worth it at this time
2. Modify the serialization logic in the yaml package to remove this field, which feels hacky
3. Set the field to a semi-realistic value, acknowledging that it's unused and proceed with the current implementation.

With those options laid out, I'm inclined to proceed with option 3, and also creating a ticket under the CAPC repo to remove this unused ObjectMeta resource.

However, one thing I still don't understand is why the K8s server accepts the top-level creationTimestamp as null but not this lower-level one. The types are defined the same way in https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/blob/main/api/v1beta2/cloudstackmachinetemplate_types.go

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:81]: When we generate the template name for cluster creation, we use `common.CPMachineTemplateName`. 

When we generate the template name for cluster Upgrade, we call `provider.getControlPlaneNameForCAPISpecUpgrade` which  checks if a new control plane template is required. If a new one is required by validating the previous EKS-A cluster spec, previous bundles, or AnyImmutableFieldChanged. If a new template is needed, it uses the same function as above, otherwise we use the existing one from the cp CloudStackMachineTemplate. 

The controller currently checks if any immutable field was changed on the CloudStackDatacenterConfig or CloudStackMachineConfigs and generates a new name if things have changed - otherwise it uses the one from the KCP in the cluster, at `cp.Spec.MachineTemplate.InfrastructureRef.Name`.

These are all sightly different ways to generate the name which make me nervous to just generate the name inside this function. It would also require feeding in a lot of context about the rest of the system to be able to check if a new one is needed or not.

File: pkg/providers/cloudstack/apibuilder.go
Comment (chrisdoherty4) [reply pkg/providers/cloudstack/apibuilder.go:96]: `kubectl` always outputs with a null creation timestamp. I've never checked if it was intentional or not but I'd imagine there's something akin to a missing `"omitempty"` somewhere.

```
~/Workspace/github.com/aws/eks-anywhere > kubectl create --dry-run=client -o yaml deployment --image=nginx nginx
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: nginx
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        resources: {}
status: {}
```

File: pkg/providers/cloudstack/apibuilder.go
Comment (chrisdoherty4) [inline pkg/providers/cloudstack/apibuilder.go:81]: Can we drop the `now` parameter? I understand the testability aspect but it puts a burden on real consumers to need to care about a time function.

File: pkg/providers/cloudstack/apibuilder.go
Comment (chrisdoherty4) [inline pkg/providers/cloudstack/apibuilder.go:17]: I think this is only used for the time func: in any case, import alias' are subject to the same rules of package names, they should be all lower case (for future reference).

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:96]: @maxdrib could we add a default for this field in the mutation webhook?

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:81]: I understand that concern. But even if we don't generate the name in this method, is possible to have a shared codepath generate the name for both controller and CLI?

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: I tried adding a defaulter to the CAPC webhook but it didn't even invoke the webhook - I think the request gets rejected before it reaches the webhooks by the K8s server

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: I think I figured it out. I found the exact same issue described in https://github.com/bitnami-labs/sealed-secrets/issues/971, and saw the fix in https://github.com/bitnami-labs/sealed-secrets/pull/979/files, which was to make the field nullable. This will require a change in CAPC

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:81]: I think sharing name generation logic makes sense, but I don't think it could be handled in this PR, which is intended to help make migration to cluster lifecycle api easier. I will create an issue for this

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:96]: I think I figured it out. It was the same issue as https://github.com/bitnami-labs/sealed-secrets/issues/971. I've created a PR in CAPC to make the field nullable: https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/pull/176, so i think we can proceed without the objectmeta being set here.

File: pkg/providers/cloudstack/cloudstack.go
Comment (maxdrib) [reply pkg/providers/cloudstack/cloudstack.go:666]: Looks like it does, thanks for the tip

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:81]: done

File: pkg/providers/cloudstack/apibuilder.go
Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go]: Fixed in the latest CAPC bundle

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [inline pkg/providers/cloudstack/apibuilder.go:40]: nit:
```suggestion
func generateMachineTemplateAnnotations(machineConfig *v1alpha1.CloudStackMachineConfigSpec) map[string]string {
```

File: pkg/providers/cloudstack/apibuilder.go
Comment (g-gaston) [inline pkg/providers/cloudstack/apibuilder.go:81]: curious why pass the `CloudStackMachineConfigSpec` instead of the `CloudStackMachineConfig`?

Comment (maxdrib) [reply pkg/providers/cloudstack/apibuilder.go:81]: The main reason was encapsulation. There's no need for this method to have access to the whole CloudStackMachineConfig - it only needs the spec

Comment (g-gaston) [reply pkg/providers/cloudstack/apibuilder.go:81]: I see
I prefer to pass the machine templates for consistency. The spec itself misses some context, it doesn't really represent anything by itself. I understand not passing big things like the cluster.Spec, but in this case it seems we lose more than we win.

That said, I don't have big concerns, I'll leave it up to you, lgtm

**Review by vivek-koppuru — APPROVED** [review]
/approve

**Pull request merged**
